### PR TITLE
Add message right click context menu v2

### PR DESCRIPTION
**Pull request by SimonBrandner**
Requires: https://github.com/vector-im/element-web/pull/17777
Fixes [#16480](https://github.com/vector-im/element-web/issues/16480)
Replaces [#5489](https://github.com/matrix-org/matrix-react-sdk/pull/5489)

<hr></hr>

Thanks to @weeman1337 for working on [#5671](https://github.com/matrix-org/matrix-react-sdk/pull/5671)!

<hr></hr>

https://user-images.githubusercontent.com/25768714/127780374-e1771641-a075-492a-b86c-6b3dc142b8b3.mp4

<hr></hr>

**For the live preview, this is always enabled but before merging this a line should be changed, so that this only works on Element Desktop and the native browser menu is used for Element Web**

Preview: https://pr5672--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/BasePlatform.ts
Comment (SimonBrandner) [inline src/BasePlatform.ts:146]: **This needs to be changed before merging!**

**Comment by t3chguy** [comment]
+ This changes existing text right click behaviour

Today: right click a word will auto-select it and let you quickly copy it
With this change: just opens the ctx menu

+ It prevents you from copying/saving inline OR `m.image` media

Opens the new ctx menu which doesn't have options of save/copy for the image you right clicked.
For inline this is especially bad as those are the only way to access those.

+ It prevents right click copy on the timestamp, this is a very desirable quick shortcut to copy link to message without having to go through

1. right click
2. click share
3. click copy
4. close dialog

(twice as many steps)

+ It prevents right click address on pills too

**Comment by SimonBrandner** [comment]
> Today: right-click a word will auto-select it and let you quickly copy it
> With this change: just opens the ctx menu

This seems to be a MacOS thing, it doesn't work for me on Linux, nor on Windows

The rest should be fixable:

+ [ ] Right-click to copy/save images
+ [ ] Right-click the timestamp to copy
+ [ ] Right-click to copy an address from a pill (it might not be a bad idea to have a separate pill context menu)

**Comment by t3chguy** [comment]
+ When right clicking input field when editing a message I am shown the ctx menu, when a word is highlighted il'll rightfully show me Copy but not Paste / Paste & Match Style

**Comment by SimonBrandner** [comment]
>     * When right clicking input field when editing a message I am shown the ctx menu, when a word is highlighted il'll rightfully show me Copy but not Paste / Paste & Match Style

Added to the list above

**Comment by ShadowJonathan** [comment]
In the original video i see that with opening the menu via the side-button, the original message stays "highlighted", but with right-clicking, it doesnt. Could it be made so that with right-clicking the message, the message in question also stays highlighted? (That behaviour is on par with discord, fwiw)

**Comment by SimonBrandner** [comment]
> In the original video i see that with opening the menu via the side-button, the original message stays "highlighted", but with right-clicking, it doesnt. Could it be made so that with right-clicking the message, the message in question also stays highlighted? (That behaviour is on par with discord, fwiw)

Done

File: src/BasePlatform.ts
Comment (turt2live) [reply src/BasePlatform.ts:146]: 🚨🚨🚨

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (turt2live) [inline src/components/views/context_menus/MessageContextMenu.tsx:50]: `export` from the other place rather than copy/paste

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (turt2live) [inline src/components/views/context_menus/MessageContextMenu.tsx:288]: this will need to use the new `mediaEventHelper` stuff, for caching purposes. Also, we can't safely have the anchor within the scope here: it would need to be done like the action bar download button (which probably means factoring it out from there into somewhere both places can use it)

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (turt2live) [inline src/components/views/context_menus/MessageContextMenu.tsx:335]: in theory if the permalinkCreator has context of the room then it should be able to pick servers for you. At least that's how I remember it working...

File: src/components/views/rooms/EventTile.tsx
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:839]: we shouldn't be using `any` - there must be a type for this.

File: src/components/views/elements/AccessibleButton.tsx
Comment (turt2live) [inline src/components/views/elements/AccessibleButton.tsx:43]: this change is a bit worrisome: the point of using `onClick` as a property is to ensure that the component can control it appropriately. If another component were to bypass this, then the component is considered to have lost control of its click handling.

A new prop for `triggerOnMouseDown = true` is ugly but maintains the component's desired capability.

File: src/components/views/rooms/EventTile.tsx
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:1066]: There's enough conditions and types in here that `!!` on the non-bools might add subconscious clarity, but not required.

**Review by turt2live — COMMENTED** [review]
overall the changes look largely fine, just a couple small things to work through. I've mentioned the download button from the action bar because that appeared after this, but landed first, so unfortunately by the laws of development branches it means this PR has to catch up 😛 (sorry). 

I'm also a bit concerned about the shear number of options in the context menu leading to a stack of variables, but don't have many immediate answers. Design might have ideas for reducing the option load.

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (SimonBrandner) [reply src/components/views/context_menus/MessageContextMenu.tsx]: That seems to be the case, this comment was copied from somewhere, though I don't remember where that was

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (SimonBrandner) [reply src/components/views/context_menus/MessageContextMenu.tsx]: In the end, I decided to drop the download option, since we already use a native menu for images and I am not entirely sure what the plan is with the download functionality. I am happy to reimplement it properly if design wants me to do so

**Review by turt2live — APPROVED** [review]

**Comment by DoM1niC** [comment]
I use this into my Fork :) I like this Contextmenu Stuff I hope this will be a part in the official Element React Code 👍🏻 @SimonBrandner can you add a point to copy a full message without selecting it first?

**Comment by SimonBrandner** [comment]
> I use this into my Fork :) I like this Contextmenu Stuff I hope this will be a part in the official Element React Code 👍🏻 

Thanks! :)

> @SimonBrandner can you add a point to copy a full message without selecting it first?

TBCH, I don't think this aligns with what we would do for Element. This is a question for design, I am sure they'll consider it

**Comment by DoM1niC** [comment]
I use your great right click PR in my Fork :) can you solve the current changes from development here ?

**Comment by DoM1niC** [comment]
new Type changes :( incoming and need to solve 💯

**Comment by janogarcia** [comment]
@SimonBrandner @niquewoodhouse This is a really great idea. Thanks, Simon, for leading the work on it! 👏 Let me add a bit of context to how we might apply it:

Web apps in general, at least when run in the browser, are usually super cautious about overriding the default behavior for the right-click interaction, respecting the platform's contextual menu unless there's a solid reason to override it.

For example, Gmail will override the context menu when right-clicking a message in the Inbox view to display the contextual actions, but it will respect it if right-clicking on a message when viewing an email thread. 

I really liked Discord desktop app's right-click context menu on messages. They don't use it, though, when Discord is run in the browser. Most probably for not messing with default browser interactions and any dependent browser plugin the user may have installed (or maybe even some assistive technologies, but I haven't researched that further).

It's also worth noting that none of the apps in the similar space are doing it either (e.g., Slack, Mattermost, Rocket.Chat, Twist, Threads), probably for the same reasons.

🔴 @niquewoodhouse My suggestion would be to decide first if we want this to happen only when in the context of the Desktop app (a sandboxed, standalone environment with different expectations about the default interactions) or in any context (a bit more invasive and potentially breaking expected default interaction patterns for users).

**Comment by ashkitten** [comment]
@janogarcia i'd like to say that as a user of element-web in the browser i've been very much looking forward to this feature. on my ultrawide monitor it's always a pain to have to figure out which controls on the right side of the screen correspond to a message all the way on the left side, and i think putting it in a context menu would be hugely beneficial to my use case.

i don't really know about all this default interactions stuff, but i will make note that things like youtube's special context menu can be bypassed by right-clicking twice or holding shift while right clicking. if that's still too surprising for web users, maybe it could be a default-off option in the settings?

**Comment by janogarcia** [comment]
@ashkitten Thanks for sharing your use case, and the note on YouTube. 👍 While completely relevant for ultra-widescreen users, it's not the case for the vast majority of users, and we need to optimize for that scenario.

Considering that we end up implementing this as the default option for the desktop app experience, we may consider adding that as a user preference. Not a fan of bloating the app with tons of settings, but this one sounds like a reasonable candidate.

/cc @niquewoodhouse

**Comment by SimonBrandner** [comment]
@janogarcia, thank you for taking a look!

Currently, the PR is designed in a way that before merging I'll disable the right-click menu for the browser, it should be one line of code. Currently, it's enabled so that the live preview can be easily tested. 

I would suggest merging this with the right-click menu only enabled on desktop and then experimenting with it in the browser if we want to

**Comment by SimonBrandner** [comment]
@janogarcia, thanks!

@turt2live, we seem to be getting close to this being approved by design/product, so I am requesting another code review as the code has changed quite a bit since the last one. Thanks!

**Comment by janogarcia** [comment]
@turt2live I'd add, please make sure it's also handling any necessary accessibility-related attribute. It seems to be managing a lot of stuff already (focus, blocks interaction/scroll with the background, keyboard navigation...).

**Comment by ashkitten** [comment]
fyi, a recent discord update seems to have enabled the existing right click context menu functionality on web.

**Comment by janogarcia** [comment]
@ashkitten Interesting. Thanks for sharing, @ashkitten. Where did you read about it? Have you seen it working already? I'm only seeing it for desktop so far.

**Comment by ashkitten** [comment]
@janogarcia i am seeing it right now with discord in firefox. maybe they're a/b testing if it's not rolled out to all users?

**Comment by janogarcia** [comment]
@ashkitten Yeah, probably a/b testing. Could you please test if they're revealing the native menu on a second right-click (as the YouTube example)?

**Comment by ashkitten** [comment]
@janogarcia discord does not reveal the native menu on a second right-click, but shift+right-click does bring it up.

**Comment by janogarcia** [comment]
@ashkitten Thanks. I'm now getting the menu on Firefox too. 👍

**Comment by kittykat** [comment]
@SimonBrandner when testing the netlify build, I found that highlighting a word and right clicking doesn't give me the same context menu as the video in the first comment shows. Is this intentional?

**Comment by janogarcia** [comment]
@kittykat The video may be outdated. That said, the example on the video you're referring to is using a deleted/redacted event as an example, which has fewer contextual actions than regular messages.

You should be getting the same actions you get on the regular context actions for that message tile, plus the `Copy` button on the top when selecting some text.

**Comment by kittykat** [comment]
@mnaturel Please stay with your original proposal for now (the back button goes back screen by screen along the whole path) and we can review the interaction when we're not in a rush as we would ideally do some user testing around it because it's a big change.

**Comment by SimonBrandner** [comment]
Here's a list of the context menu buttons with info about when they're visible:

- Native options
    - Copy button - when there is selected text
- Quick options - only visible in right-click context menu, otherwise handled by the message action bar
    - React - when user can react to an event
    - Reply - When user can reply to an event
    - Edit - when user can edit event
- Common options - these should behave the same as they do in the`Options` action bar dropdown
    - View in room - if the event is a root of a thread
    - Open in map site - when the event is a location event and user can view in a map viewing app
    - End poll - when the event is poll event that hasn't been ended yet and user can end
    - Quote - when event is a textual event
    - Forward - if user can forward (event is not location event)
    - Pin button - when user has permission to pin/unpin the event
    - Permalink - when we can get a permalink for the event (probably should be always, probably just a check if some code went awry)
    - Report - when the user isn't "me"
    - External URL - something to do with bridges
    - Show preview - something to do with inline widgets?
    - View source - always
    - Resend reactions - when the event hasn't been redacted and there are reacting that failed to send
    - Collapse reply chain - when there is an expanded reply chain
- Redact options
    - Redact - when user can redact

**Comment by janogarcia** [comment]
@SimonBrandner Just a quick note that @niquewoodhouse and I are actively reviewing the implications of right-click context menus across every possible context in the timeline.

screencapture-github-matrix-org-design-issues-231-2022-03-04-10_41_56

**Comment by turt2live** [comment]
@janogarcia does that plan mean this PR is likely to change? Wondering if it's in a position for code review, and subsequent merging, or if we should hold off a bit.

**Comment by janogarcia** [comment]
@turt2live We’re actively working on it, making progress with the tasks above. We definitely want this functionality to get added/merged. It’s just a matter of figuring out every relevant case before getting it pushed to production.

That said, given that it may require a considerable effort, we don’t want this to be on Simon’s plate just because he kindly initiated the work on it (it would be very welcome if he _can/wants to_ keep contributing to it). So what would be the best way to handle this?

Maybe merge it into develop and then keep working/iterating on it? Unsure how we handle that scenario where we’re interested in merging a PR, but the scope for the proposed change gets more complex.

---

Here's some examples notes I made:

Web and Desktop- Right-click context menus · Issue #231 · matrix-org-design

**Comment by SimonBrandner** [comment]
> That said, given that it may require a considerable effort, we don’t want this to be on Simon’s plate just because he kindly initiated the work on it (it would be very welcome if he _can/wants to_ keep contributing to it). 

I am happy to keep working on the PR if it fits my schedule

> So what would be the best way to handle this? Maybe merge it into develop and then keep working/iterating on it? 

If you're ok with the PR in its current form, I would suggest merging it and I am happy to iterate on it in other PRs. We could probably also hide it behind a labs flag if necessary (but I am not sure if it is necessary). 

That said I am not against other options, let me know what you prefer

**Comment by turt2live** [comment]
Usually in scenarios like this we will try to find team bandwidth somewhere if the contributor is unable/unwilling (which thankfully doesn't look to be the case here). 

Indeed, if this change is suitable for develop as-is then I'd recommend we land it, open some issues, and kindly ask Simon to pick them up if free & able.

**Comment by janogarcia** [comment]
@turt2live Makes sense. 👍 Thanks for clarifying! Glad to hear that there’s a way for us to unblock this PR and then keep working on improving it. Sorry Simon, for not being familiar with this workflow before!

To minimize the confusion on develop I’d highly recommend that we **hide it behind a Labs flag** as @SimonBrandner suggested. Would you be willing to do that before merging?

**Comment by SimonBrandner** [comment]
> To minimize the confusion on develop I’d highly recommend that we **hide it behind a Labs flag** as @SimonBrandner suggested. Would you be willing to do that before merging?

Yes, sure will have a look into that!

**Comment by SimonBrandner** [comment]
> > To minimize the confusion on develop I’d highly recommend that we **hide it behind a Labs flag** as @SimonBrandner suggested. Would you be willing to do that before merging?
> 
> Yes, sure will have a look into that!

Done, it is a labs flag now! See https://github.com/vector-im/element-web/pull/21480 for docs

**Comment by turt2live** [comment]
just a note to say I haven't forgotten about this, just that I'm waiting for the removal of skinning to land which might affect the code's approach here. Also, we recently moved View Source behind developer options, so if that hasn't been reflected here then please do 😇

**Comment by SimonBrandner** [comment]
> Also, we recently moved View Source behind developer options, so if that hasn't been reflected here then please do innocent

It should work as expected

**Comment by SimonBrandner** [comment]
(assuming e2e tests are failing due to unrelated reasons as they pass locally)

File: src/BasePlatform.ts
Comment (turt2live) [reply src/BasePlatform.ts:152]: @SimonBrandner does this still need changing? If so, can we make that change for merging?

File: src/components/structures/ContextMenu.tsx
Comment (turt2live) [inline src/components/structures/ContextMenu.tsx:435]: we should be keeping the return type

File: src/components/structures/ContextMenu.tsx
Comment (turt2live) [inline src/components/structures/ContextMenu.tsx:432]: if we want a partial, we should define what properties we actually want instead (so that DOMRect automatically fulfills that interface).

File: src/components/views/elements/AccessibleButton.tsx
Comment (turt2live) [inline src/components/views/elements/AccessibleButton.tsx:44]: err, why All? It's a button...

File: src/components/views/rooms/EventTile.tsx
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:232]: if this can be null, it should be wrapped in `Optional` or similar

File: src/components/views/rooms/EventTile.tsx
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:941]: can we move the render function to near the other render function please?

File: src/components/views/rooms/EventTile.tsx
Comment (turt2live) [inline src/components/views/rooms/EventTile.tsx:968]: ideally we return `null` when there's no menu rather than an empty fragment, for a slight performance boost.

**Review by turt2live — CHANGES_REQUESTED** [review]
generally looks good to me, thanks!

Will be removing design and product reviews because of labs flag.

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (turt2live) [inline src/components/views/context_menus/MessageContextMenu.tsx]: these are important for middle click and other anchor semantics

**Review by turt2live — COMMENTED** [review]
almost there :D

File: src/components/views/context_menus/MessageContextMenu.tsx
Comment (SimonBrandner) [reply src/components/views/context_menus/MessageContextMenu.tsx]: I was thinking there might be a way to use generics here but I don't think there would be a way to avoid the duplicity of the `element` prop and the type

Comment (turt2live) [reply src/components/views/context_menus/MessageContextMenu.tsx]: I'm a bit confused as to how this was working before - did we convert something to TypeScript that sits between here and the AccessibleButton?

Extending all the types is not great - if there's a type out there that determines the attributes based on the tag name then we can use that, otherwise I guess we're left with no choice :(

Comment (SimonBrandner) [reply src/components/views/context_menus/MessageContextMenu.tsx]: I am not sure what you mean by before - the copy link button is new, this didn't work before.

Yeah, I am afraid our only option is to use all :(

Comment (turt2live) [reply src/components/views/context_menus/MessageContextMenu.tsx]: the code was there previously, but it's been removed/altered by the PR, so it did work at some point.

It'd be the comment below which made it work, at a glance:
```
// XXX: Typescript signature for AccessibleButton doesn't work properly for non-inputs like `a`
```

Comment (SimonBrandner) [reply src/components/views/context_menus/MessageContextMenu.tsx]: Oh, I see, should be fixed now! Sorry, for not seeing this earlier :sweat_smile:

**Review by turt2live — APPROVED** [review]
:shipit:

**Pull request merged**
